src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+                Set<String> microarrayIdentifierLines = new HashSet<>();
+                for (String microarrayResponseLine : microarrayResponseLines) {
+                    List<String> tabSplit = Arrays.asList(microarrayResponseLine.split("\t"));
+                    // Microarray ID is 4th column, and sometimes it isn't in a line.
+                    // Since we need the ID, we require the List to have 4 elements.
+                    if (tabSplit.size() == 4) {
+                        microarrayIdentifierLines.add(microarrayResponseLine + "\n");
+                    }
+                }
+
+                // Write data to file.
+                if (!microarrayIdentifierLines.isEmpty()) {
+                    File speciesMicroarrayFile = createNewFile(this.destination + biomartSpeciesName + "_microarray");
+                    for (String probeIdLine : microarrayIdentifierLines) {
+                        Files.write(Paths.get(speciesMicroarrayFile.toString()), probeIdLine.getBytes(), StandardOpenOption.APPEND);
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Attempts to create a connection to Biomart with the supplied query
+     * @param queryString String of Biomart query
+     * @return Set<String> Biomart response data.
+     * @throws Exception Can be caused by the query URL not being valid, or the data not existing for the species.
+     */
+    private Set<String> queryBiomart(String queryString, int retryCount) throws Exception {
+        // Create connection to Biomart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == 200) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                // Biomart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    retryCount++;
+                    if (retryCount < 5) {
+                        Thread.sleep(10000);
+                        logger.warn("Biomart query failed. Trying again...");
+                        return queryBiomart(queryString, retryCount);
+                    }
+                    throw new Exception("Biomart query failed with message: " + line);
```

Comment:
Try not to use generic exceptions whenever possible.  It's best practice to use an existing standard exception that is a subclass of `Exception` or to create a custom exception class you write yourself that inherits from `Exception` when needed (though using an existing standard `Exception` child class is preferred).  

In this case, I don't know of an appropriate standard exception.  You may want to create your own extending `Exception` if you don't want it to be an unchecked Exception like RuntimeException that ends the program.

For reference, I found a list of Java Exceptions that might be handy for selecting standard ones to use: https://programming.guide/java/list-of-java-exceptions.html

Reply:
Once you've introduced standard or custom exceptions in place of `Exception`, be sure to change the method signature to throw those specific exceptions and document them with each having their own `@throws` tag in the JavaDocs for the method.  Then you can document under what circumstances each kind of exception is thrown.

For this method, for example, you would have:

```
[...other JavaDoc annotations]
@throws InterruptedException Thrown if the sleep delay between BioMart query re-attempts is interrupted
@throws IOException Thrown if unable to parse the response from the BioMart URL
@throws HTTPException Thrown if unable to successfully connect (i.e. a response of 200 OK) to BioMart after retrying queries a predetermined maximum number of times
private Set<String> queryBiomart(String queryString, int retryCount) throws InterruptedException, IOException, HttpException {
```

You can add your custom exception for a query failure with a 200 response returning a result with the "ERROR" string to the method declaration and @throws list when you create it.

Reply:
I was thinking for the custom exception you could do something like this:

```
private class BioMartQueryException extends Exception {
    BioMartQueryException(String message) {
        super("BioMart query failed with message: " + message);
    }
}
```

Reply:
Ok I've added in this custom exception, thanks for your help. I'll be updating the JavaDocs soon